Marlin/Marlin_main.cpp

```diff
@@ -284,11 +284,11 @@ bool axis_homed[3] = { false };
 
 static long gcode_N, gcode_LastN, Stopped_gcode_LastN = 0;
 
-static char* current_command, *current_command_args;
-static int cmd_queue_index_r = 0;
-static int cmd_queue_index_w = 0;
-static int commands_in_queue = 0;
 static char command_queue[BUFSIZE][MAX_CMD_SIZE];
+static char* current_command, *current_command_args;
+static uint8_t cmd_queue_index_r = 0,
+               cmd_queue_index_w = 0,
+               commands_in_queue = 0;
```

Comment:
Sure enough, the variable is treated as if it were `volatile` in the `loop()` function…
```asm
    547a:	80 91 a5 04 	lds	r24, 0x04A5
ExitUnknownCommand:

  // Still unknown command? Throw an error
  if (!code_is_good) unknown_command_error();

  ok_to_send();
    547e:	88 23       	and	r24, r24

  #if ENABLED(SDSUPPORT)
    card.checkautostart(false);
  #endif

  if (commands_in_queue) {
    5480:	81 f0       	breq	.+32     	; 0x54a2 <loop+0x34>
      else
        process_next_command();

    #else

      process_next_command();
    5482:	85 dc       	rcall	.-1782   	; 0x4d8e <_Z20process_next_commandv>
    5484:	80 91 a5 04 	lds	r24, 0x04A5

    #endif // SDSUPPORT

    // The queue may be reset by a command handler or by code invoked by idle() within a handler
    if (commands_in_queue) {
    5488:	88 23       	and	r24, r24
    548a:	59 f0       	breq	.+22     	; 0x54a2 <loop+0x34>
    548c:	81 50       	subi	r24, 0x01	; 1
      --commands_in_queue;
    548e:	80 93 a5 04 	sts	0x04A5, r24
    5492:	80 91 a7 04 	lds	r24, 0x04A7
      cmd_queue_index_r = (cmd_queue_index_r + 1) % BUFSIZE;
    5496:	90 e0       	ldi	r25, 0x00	; 0
    5498:	01 96       	adiw	r24, 0x01	; 1
    549a:	83 70       	andi	r24, 0x03	; 3
    549c:	99 27       	eor	r25, r25
    549e:	80 93 a7 04 	sts	0x04A7, r24
```
See that immediately after `process_next_command()` is called, `commands_in_queue` is reloaded from RAM: `lds r24, 0x04A5`. So the compiler is smartly not assuming it was unaltered.